File: Measurement Studio/Examples/DotNET/v4.5/UI.WinForms/Graph/ErrorBands/CustomErrorData/vb/CombinationErrorMode.vb

Imports System
Imports NationalInstruments.UI
Imports NationalInstruments.UI.WindowsForms

Public Class CombinationErrorMode
    Inherits XYErrorDataMode
    Private ReadOnly constantOffset As Double
    Private ReadOnly percentOffset As Double
    Private ReadOnly threshold As Integer

    Public Sub New(ByVal constantOffset As Double, ByVal percentOffset As Double, ByVal threshold As Integer)
        Me.constantOffset = constantOffset
        Me.percentOffset = percentOffset
        Me.threshold = threshold
    End Sub

    Public Overrides Sub GetErrorData(ByVal args As XYErrorDataArgs, ByRef highErrorData() As Double, ByRef lowErrorData() As Double)
        Dim data() As Double
        Dim xData() As Double = args.GetXData()
        Select Case args.PrimaryErrorDataType
            Case XYDataType.XData
                data = xData

            Case XYDataType.YData
                data = args.GetYData()

            Case Else
                data = New Double(0) {}
        End Select

        highErrorData = New Double(data.Length - 1) {}
        lowErrorData = New Double(data.Length - 1) {}
        Dim offset As Double
        Dim i As Integer
        For i = 0 To data.Length - 1
            If xData(i) > threshold Then
                ' If the x value of a data point goes over the threshold, use a percentage.
                offset = data(i) * percentOffset
            Else
                ' Otherwise, use a constant value.
                offset = constantOffset
            End If

            highErrorData(i) = offset
            lowErrorData(i) = offset
        Next
    End Sub

End Class

File: Measurement Studio/Examples/DotNET/v4.5/UI.WinForms/Graph/ErrorBands/CustomErrorData/cs/CombinationErrorMode.cs

using System;
using NationalInstruments.UI;
using NationalInstruments.UI.WindowsForms;

namespace NationalInstruments.Examples.CustomErrorData
{

	public class CombinationErrorMode : XYErrorDataMode
	{
        private readonly double constantOffset;
        private readonly double percentOffset;
        private readonly int threshold;

		public CombinationErrorMode(double constantOffset, double percentOffset, int threshold)
		{
            this.constantOffset = constantOffset;
            this.percentOffset = percentOffset;
			this.threshold = threshold;
		}

		public override void GetErrorData(XYErrorDataArgs args, out double[] highErrorData, out double[] lowErrorData)
		{
			double[] data;
            double[] xData = args.GetXData();
            switch(args.PrimaryErrorDataType)
			{
				case XYDataType.XData:
					data = xData;
					break;

				case XYDataType.YData:
					data = args.GetYData();
					break;

				default:
					data = new double[0];
					break;
			}

			highErrorData = new double[data.Length];
			lowErrorData = new double[data.Length];
            double offset;
            for(int i = 0; i < data.Length; ++i)
			{
				// If the x value of a data point goes over the threshold, use a percentage.
				if(xData[i] > threshold)
				{
                    offset = data[i] * (percentOffset / 100);
				}
				// Otherwise, use a constant value.
				else
				{
					offset = constantOffset;
				}

                highErrorData[i] = offset;
                lowErrorData[i] = offset;
			}
		}

	}

}
